### ISSUE TITLE
Mixing `shake` with `optparse-applicative`  command line options

### ISSUE THREAD

**Issue by Boarders** [ndmitchell/shake#824]
I am trying to write a basic shake file for building C projects but wanted to add some command line options using `optparse-applicative`. Here is my draft (based on the shake tutorial build script):

```haskell
module Main where

import Development.Shake
import Development.Shake.FilePath
import Options.Applicative
import Data.Foldable
import qualified System.Directory as Dir
import Control.Exception (bracket)

main :: IO ()
main = do
  BuildOptions
    { projDir = proj
    , srcDir = src
    , incDir = inc
    , buildDir = build
    } <- execParser parseOpts
  bracket
    (do
        currDir <- Dir.getCurrentDirectory
        Dir.setCurrentDirectory (currDir </> proj)
        pure currDir
    )
    (\currDir -> Dir.setCurrentDirectory currDir)
    (\_ ->
       shakeArgs shakeOptions $ do
         want [build </> "run" <.> exe]
     
         phony "clean" $ do
             putInfo "Cleaning files in build directory"
             removeFilesAfter build ["//*"]
     
         build </> "run" <.> exe %> \out -> do
           cs <- getDirectoryFiles src ["//*.c"]
           liftIO $ Dir.createDirectoryIfMissing True build
           let os = [build </> c -<.> "o" | c <- cs]
           need os
           cmd_ "gcc -o" [out] "-I" inc os
     
         build </> "*.o" %> \out -> do
           let c = dropDirectory1 $ out -<.> "c"
           let m = out -<.> "m"
           cmd_ "gcc -c" [c] "-o" [out] "-MMD -MF" [m]
    )
       where
    parseOpts =
      info (parseBuildOptions <**> helper) $
        fold
          [ fullDesc
          , progDesc "A shake file for building C projects"
          , header "shake-C: build your C project"
          ]

parseBuildOptions :: Parser BuildOptions
parseBuildOptions =
  BuildOptions
  <$> strOption
       (fold
         [ long "proj"
         , metavar "PROJECT DIRECTORY"
         , value "."
         , help "project directory *.c files"
         ]
       )
  <*> strOption
       (fold
         [ long "src"
         , metavar "SOURCE DIRECTORY"
         , value "src"
         , help "source directory containing *.c files"
         ]
       )
  <*> strOption
       (fold
         [ long "inc"
         , metavar "INCLUDE DIRECTORY"
         , value "inc"
         , help "include directory containing *.h files"
         ]
       )     
  <*> strOption
       (fold
         [ long "build"
         , metavar "BUILD DIRECTORY"
         , value "_build"
         , help "build directory containing output of build"
         ]
       )       

data BuildOptions = BuildOptions
  { projDir :: FilePath
  , srcDir   :: FilePath
  , incDir   :: FilePath
  , buildDir :: FilePath
  }
  ```
  
  When I try to run this I get the following outputs:
  ```shell
  cabal run exe -- --proj
Up to date
The option `--proj` expects an argument.

Usage: exe [--proj PROJECT DIRECTORY] [--src SOURCE DIRECTORY] 
           [--inc INCLUDE DIRECTORY] [--build BUILD DIRECTORY]

  A shake file for building C projects
  ```
  
  ```shell
  cabal run exe -- --proj c-proj

p to date
shake: unrecognized option `--proj'

Usage: exe [options] [target] ...

Options:
[lots of shake options[
```
Am I mixing these two in a way one shouldn't and if so is there a better way to be doing things?

**Comment by ndmitchell** [comment]
Shake provides a few entry points:

* `shake`, which doesn't process the arguments in any way. If you want to use optparse or any other command line library, switch to `shake`, and it should work. The issue above is that you get Shake to parse the command line arguments in addition to optparse, and they have different syntaxes.
* `shakeArgs` which parses the arguments from `getArgs`.
* `shakeArgsWith` which parses the arguments from `getArgs` and also allows you to add extra flags. That would be an alternative to using optparse (simpler, but less powerful, so optparse is still useful in cases)

TLDR: s/shakeArgs/shake/ above should fix it

**Comment by Boarders** [comment]
Perfect, thank you! Sorry for the pedestrian question.

**Comment by ndmitchell** [comment]
No worries, I can see that the behaviour of using the command line twice is super confusing!
